SMQ-3173 - Use `buildx` for Multi-Platform Docker Image Support

rodneyosodo opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `SMQ-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/supermq/blob/main/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

This is a feature because it adds riscv support

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

Adds riscv support
<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

- Resolves #3172 

## Have you included tests for your changes?

Yes, tested manually https://hub.docker.com/r/rodneydav/supermq-users/tags

<img width="1460" height="628" alt="image" src="https://github.com/user-attachments/assets/334e2a31-f176-4158-8a0e-6e7dee7bda94" />

<!--If you have not included tests, please explain why.
For example:
Yes, I have included tests for my changes.
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

rodneyosodo committed:
fix(docker): prefix images with docker.io registry

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(docker): update go version and refine certificate copy

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
feat(docker): add platform support for docker buildx

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(ci): remove fetch tags with checkout options

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix: revert testing changes

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
chore(docker): upgrade postgres docker image

This image supports riscv and arm architectures

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
chore(docker): update redis image

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
chore(docker): update nginx image

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
chore(docker): update jaeger image

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
chore(docker): update rabbitmq and nats images

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(ci): remove cd on pull request

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on docker/Dockerfile.dev:
```diff
@@ -1,8 +1,10 @@
 ARG SVC
 COPY $SVC /exe
-COPY --from=alpine:latest /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+COPY --from=alpine /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
```
No need for naming alpine stage (`FROM alpine:3.22 AS alpine`), we can use `COPY --from=alpine:3.22 /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt`

dborovcanin commented on .github/workflows/build.yaml:
```diff
@@ -16,12 +16,22 @@ jobs:
       contents: read
 
     steps:
+      - name: Free Disk Space (Ubuntu)
```
Why is  this needed?

rodneyosodo replied:
This is because we run out of space when building the images. I had initially started with a larger pool `darwin/amd64,darwin/arm64,freebsd/amd64,freebsd/arm64,netbsd/amd64,netbsd/arm64,openbsd/amd64,openbsd/arm64,linux/amd64,linux/arm64,linux/riscv64,windows/amd64,windows/arm64`, this is when I experienced it

dborovcanin commented on Makefile:
```diff
@@ -52,10 +52,9 @@ define make_docker
 	$(eval svc=$(subst docker_,,$(1)))
 
 	docker build \
+		--platform linux/amd64,linux/arm64,linux/riscv64 \
```
Let's switch to buildx.

rodneyosodo committed:
fix: revert to COPY --from=remote image

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(docker): change to buildx

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin reviewed (changes requested):
Let's create separate build tags for RISC V and ARM, and have them outside of the regular CI push so we don't waste our CI time on it.

rodneyosodo committed:
fix: only build for linux/amd64 images

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on Makefile:
```diff
@@ -51,11 +51,10 @@ endef
 
-	docker build \
+	docker buildx build --load \
+		--platform linux/amd64 \
```
Let's add a Makefile target for building and tagging dockers for different architecture.

rodneyosodo committed:
feat(docker): make platforms a variable

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
revert buildx

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
chore: add command to publish docker image of different architectures

For Linux ARM64:
make publish_arch GOOS=linux GOARCH=arm64

For Linux RISC-V 64:
make publish_arch GOOS=linux GOARCH=riscv64

For Linux ARM v7:
make publish_arch GOOS=linux GOARCH=arm GOARM=7

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
